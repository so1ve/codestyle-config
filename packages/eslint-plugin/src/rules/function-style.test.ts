import type { InvalidTestCase } from "eslint-vitest-rule-tester";
import { unindent as $ } from "eslint-vitest-rule-tester";
import { expect } from "vitest";

import { run } from "./_test";
import rule, { RULE_NAME } from "./function-style";

const valid = [
	"const a = () => 1;",
	$`
	  function a() {
	    stuff;
	    return 1;
	  }
	`,
	$`
	  function a(some: Type)
	    function a(some: Another)
	    function a() {
	      return 1;
	  }
	`,
	$`
	  const a: Annotation = () => {
	    stuff;
	    return 1;
	  };
	`,
	"function* a() {}",
	"const a = () => {}",
	"function a() {}",
	// This is fucking ugly but cannot be fixed
	"const a: Type = async function foo(): Returns {}",
	"const a = () => { a = this; return 1; }",
	"function a() { return this; }",
	"function a() { return () => this }",
	"const a = () => { foo; function a() { this; } }",
	$`
	  export function last(array: readonly []): undefined;
	  export function last<T>(array: readonly T[]): T;
	  export function last<T>(array: readonly T[]): T | undefined {
	    return at(array, -1);
	  }
	`,
	$`
	    export default function () {
	    return
	  }
	`,
	$`
	  function a() {
	    return
	  }
	`,
	"const a = () => { const b = () => { return this; }; return this; };",
	$`
	    function a() {
	    // foo
	    return 1
	  }
	`,
];

const invalid: InvalidTestCase[] = [
	"const a = () => { return 1; };",
	"const a = () => { stuff; return 1; };",
	"const a = () => { stuff; };",
	"const a = async () => { return 1; };",
	"const a = async () => { stuff; return 1; };",
	"const a = async (): Promise<1> => { stuff; return 1; };",
	"const a: Some = async (): Type => { return 1; };",
	"async function a(): Type { return 1; }",
	"const a = async function foo(): Returns {}",
	$`
	  const a = () => {
	       return {
	         a: 1,
	       };
	     };
	`,
	$`
	  	export default function a() {
	    return {};
	  }
	`,
	$`
	  	export default function() {
	    return {};
	  }
	`,
	$`
	  	function foo() {
	    return [
	      // foo
	    ];
	  }
	`,
	"function foo() { return{} }",
];

run({
	name: RULE_NAME,
	rule,

	valid,
	invalid: invalid.map(
		(i): InvalidTestCase =>
			typeof i === "string"
				? {
						code: i,
						output: (o) => expect(o).toMatchSnapshot(),
					}
				: i,
	),
});
