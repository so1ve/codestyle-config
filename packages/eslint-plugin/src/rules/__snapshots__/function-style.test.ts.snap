// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`function-style > invalid > 	export default function a() {
  return {};
} 1`] = `"	export default () => ({});"`;

exports[`function-style > invalid > 	export default function() {
  return {};
} 1`] = `"	export default () => ({});"`;

exports[`function-style > invalid > 	function foo() {
  return [
    // foo
  ];
} 1`] = `
"	const foo = () => ([
    // foo
  ]);"
`;

exports[`function-style > invalid > async function a(): Type { return 1; } 1`] = `"const a = async (): Type => (1);"`;

exports[`function-style > invalid > const a = () => {
     return {
       a: 1,
     };
   }; 1`] = `
"const a = () => ({
       a: 1,
     });"
`;

exports[`function-style > invalid > const a = () => { return 1; }; 1`] = `"const a = () => (1);"`;

exports[`function-style > invalid > const a = () => { stuff; }; 1`] = `"function a() { stuff; }"`;

exports[`function-style > invalid > const a = () => { stuff; return 1; }; 1`] = `"function a() { stuff; return 1; }"`;

exports[`function-style > invalid > const a = async () => { return 1; }; 1`] = `"const a = async () => (1);"`;

exports[`function-style > invalid > const a = async () => { stuff; return 1; }; 1`] = `"async function a() { stuff; return 1; }"`;

exports[`function-style > invalid > const a = async (): Promise<1> => { stuff; return 1; }; 1`] = `"async function a(): Promise<1> { stuff; return 1; }"`;

exports[`function-style > invalid > const a = async function foo(): Returns {} 1`] = `"async function a(): Returns {}"`;

exports[`function-style > invalid > const a: Some = async (): Type => { return 1; }; 1`] = `"const a: Some = async (): Type => (1);"`;

exports[`function-style > invalid > function foo() { return{} } 1`] = `"const foo = () => ({});"`;
